crates/api_common/src/request.rs

```diff
@@ -63,71 +59,61 @@ pub async fn fetch_link_metadata(
-      .await
-      .unwrap_or_default(),
-    _ => Default::default(),
-  }
-}
 /// Generate post thumbnail in background task, because some sites can be very slow to respond.
 ///
 /// Takes a callback to generate a send activity task, so that post can be federated with metadata.
+///
+/// TODO: `federated_thumbnail` param can be removed once we federate full metadata and can
+///       write it to db directly, without calling this function.
+///       https://github.com/LemmyNet/lemmy/issues/4598
 pub fn generate_post_link_metadata(
   post: Post,
   custom_thumbnail: Option<Url>,
+  federated_thumbnail: Option<Url>,
   send_activity: impl FnOnce(Post) -> Option<SendActivityData> + Send + 'static,
   local_site: Option<LocalSite>,
   context: Data<LemmyContext>,
 ) {
   spawn_try_task(async move {
-    // Decide if the thumbnail should be generated
-    let allow_sensitive = local_site_opt_to_sensitive(&local_site);
-    let page_is_sensitive = post.nsfw;
-    let allow_generate_thumbnail = allow_sensitive || !page_is_sensitive;
-    let do_generate_thumbnail =
-      allow_generate_thumbnail && custom_thumbnail.is_none() && post.thumbnail_url.is_none();
-
-    // Generate local thumbnail only if no thumbnail was federated and 'sensitive' attributes allow it.
-    let metadata = fetch_link_metadata_opt(
-      post.url.as_ref().map(DbUrl::inner),
-      do_generate_thumbnail,
-      &context,
-    )
-    .await;
-
-    // If its an image post, it needs to overwrite the thumbnail, and take precedence
-    let image_url = if metadata
+    let metadata = match &post.url {
+      Some(url) => fetch_link_metadata(url, &context).await.unwrap_or_default(),
+      _ => Default::default(),
+    };
+
+    let is_image_post = metadata
       .content_type
       .as_ref()
-      .is_some_and(|content_type| content_type.starts_with("image"))
-    {
-      post.url.map(Into::into)
-    } else {
-      None
-    };
+      .is_some_and(|content_type| content_type.starts_with("image"));
+
+    // Decide if we are allowed to generate local thumbnail
+    let allow_sensitive = local_site_opt_to_sensitive(&local_site);
+    let allow_generate_thumbnail = allow_sensitive || !post.nsfw;
 
-    // Build the thumbnail url based on either the post image url, custom thumbnail, metadata fetch, or existing thumbnail.
-    let thumbnail_url = image_url
-      .or(custom_thumbnail)
-      .or(metadata.thumbnail.map(Into::into))
-      .or(post.thumbnail_url.map(Into::into));
+    // Use custom thumbnail if available and its not an image post
+    let thumbnail_url = if !is_image_post && custom_thumbnail.is_some() {
+      custom_thumbnail
+    }
+    // Use federated thumbnail if available
+    else if federated_thumbnail.is_some() {
+      federated_thumbnail
+    }
+    // Generate local thumbnail if allowed and post.url is Some
+    else if let (true, Some(url)) = (allow_generate_thumbnail, &post.url) {
```

Comment:
This is only if its an image post also. How about:

```rust
let thumbnail_url = if is_image_post {
  if let ....
} else {
  if custom_thumbnail.is_some() {
    ...
  } else if federated_thumbnail.is_some() { 
    ...
  }
}

Reply:
The logic seems too complicated to work with simple or chaining:
- It doesnt make sense to use post.url as thumbnail
- Custom thumbnail should only be used if its not an image post
- In some cases we need to generate a local thumbnail, in others not

Reply:
Updated to consider opengraph image for local thumbnail generation.